### DISCUSSION TITLE
Automaterialization: not all upstream assets considered for overdue freshness policy

### DISCUSSION THREAD

**Discussion by dondelicaat** [dagster-io/dagster#19187] [Q&A]
We are using automaterialization together with freshness policies in order to schedule our DBT models. We like the concept but run into quite some unexpected behaviour. We are currently only defining freshness policies on the final reports and for the rest have lazy materialization on all the assets. Now we have an issue that all the sources are up-to-date according to the freshness policy but the freshness policy does not trigger all upstream assets to be materialized. In particular I noticed that the following happens often:

We have a source asset A -> model B -> model C -> model D (with freshness policy)

now model C gets triggered in order to match D's freshness policy: 

![image](https://github.com/dagster-io/dagster/assets/15779824/5f617331-4c8c-403d-a12f-e00011e947d3)

but that one says it's waiting on asset B. The source data for asset B is fresh but asset B does not get the 'required by downstream asset' condition triggered:

![image](https://github.com/dagster-io/dagster/assets/15779824/0e50af4d-9bbc-4565-a7c8-3560daa0d8cb)

and therefore the whole chain just gets stuck and the assets never materialize. I was expecting that both assets would have the condition: "Required to meet this or downstream asset's freshness policy" checked. If I then go and manually materialize asset B, C and D then D says that it is fresh as was expected by looking at the upstream sources. 

So it seems that on every evaluation tick not all assets are considered for materialization even though that is required by the downstream freshness policy... Am I missing something or is there a good explanation for this behaviour?

**Comment by OwenKephart** [comment]
Hi @dondelicaat -- in this case, I believe you should be able to resolve this by setting your policy to:

```python
from dagster import AutoMaterializePolicy, AutoMaterializeRule

my_policy = AutoMaterializePolicy.lazy().without_rules(
    AutoMaterializeRule.skip_on_parent_outdated(),
)
```

The "Required to meet this or downstream asset's freshness policy" rule will (somewhat misleadingly) only activate if "new enough" upstream data exists in order for it to make sense to materialize that asset.

That new enough data does not exist because the upstream asset did not materialize, and the upstream asset did not materialize because, by the time it came to evaluate it, there was even newer data upstream data of an ancestor, activating that skip rule.

Definitely arguable that this skip rule shouldn't be part of the lazy policy to begin with.

**Comment by bolinzzz** [comment]
Piggybacking on @dondelicaat's question, I have the following set up
```mermaid
graph LR
  subgraph code_location_1
    a --> b
  end

  subgraph code_location_2
    c --> d
  end

  b --> c 
```

a is a daily partitoned asset
b, c and d are non partitioned assets.

a and b have eager auto-materialise policy, c has lazy auto-materialise policy. Lastly, d has lazy auto-materialise policy with the following freshness policy `FreshnessPolicy(maximum_lag_minutes=5)`

Based on my assumption, I am expecting all asset a to d to be materialised every 5 minutes in order to meet d's freshness policy. However, nothing seems to be auto-materialising. It simply states that d is overdue. 
![Screenshot 2024-01-17 at 6 05 46 PM](https://github.com/dagster-io/dagster/assets/101250079/536dcd79-9f05-43f8-9569-db1dde0b6144)


@OwenKephart could you kindly provide some guidance on how to make this work? Is it because this does not work across code locations?

**Comment by dondelicaat** [comment]
@bolinzzz You might also run into issues with using this across code locations. I raised a bug for this earlier https://github.com/dagster-io/dagster/issues/18905

@OwenKephart 
> The "Required to meet this or downstream asset's freshness policy" rule will (somewhat misleadingly) only activate if "new enough" upstream data exists in order for it to make sense to materialize that asset.
>
> That new enough data does not exist because the upstream asset did not materialize, and the upstream asset did not materialize because, by the time it came to evaluate it, there was even newer data upstream data of an ancestor, activating that skip rule.

Actually it is the other way around if I understand you correctly - namely; the asset after the root asset does not have "Required to meet this or downstream asset's freshness policy"  but the child asset of that _does_. The root asset is a partitioned asset that just gets materialized on a schedule. So after that partition exist "new enough" data is there so it should just be able to materialize the downstream asset.

It's quite hard to debug (and explain) honestly. What it boils down to is that the scheduler seems deadlocked and that the freshness policy should just trigger a rematerialization. If I just trigger the materialization manually then all freshness policies are not longer overdue.

**Reply by bolinzzz** [reply]
Thanks, @dondelicaat! Our setup closely mirrors yours: one location for code with partitioned extract and load assets that auto-materialize eagerly, and another for dbt assets handling "BI" work, auto-materializing lazily with defined freshness policies for tail assets.
